### ISSUE TITLE
Idea's on Plugins and Security

### ISSUE THREAD

**Issue by mmstick** [redox-os/ion#508]
We should probably have a `ION_PLUGINS_ENABLED` variable that is set to 1 by default, but can be set to 0 to disable support for plugins. In addition, maybe we could also implement an optional `ION_PLUGINS_METHOD_WHITELIST` and `ION_PLUGINS_NAMESPACE_WHITELIST` array which can be used to designate plugins that are explicitly allowed to execute. Then maybe we could have a means of protecting certain variables from being set within scripts, or even only allowing certain variables to be set from an environment file. Suggestions welcome for ways to make Ion more secure.

**Comment by ids1024** [comment]
Do panics in plugins crash ion currently?

**Comment by mmstick** [comment]
I'm not handling panics in any way, so probably. I've not tried.

**Comment by BafDyce** [comment]
This definitely seems like a good idea!
Right now, variables such as `HOME`, `USER`, `?` are writeable, which should never be the case.

I'd like to also use the chance to ask what exactly methods are? I guess they are neither functions, nor builtins, nor plugins?

**Comment by mmstick** [comment]
@BafDyce Correct. They are one of the key features of Ion: string and array methods. They are based on the same idea from the Oil shell that is also in development (not by us, of course). Our variant of methods were expanded to also allow in-line expressions though, in addition to supporting a default type. They are used to perform some specialized expansions during the parsing stage of the pipeline statements. The type returned by the method is denoted by the sigil, and the methods that can be called are associated with that sigil. They look like this in practice:

```ion
let string_var = "Hello Shell"

for character in @graphemes(string_var)
    echo $character
end

let word_array = [ @split(string_var) ]
for word in @word_array
    echo $word
end

let joined = $join(word_array, "-")
echo $joined
```

But we can also use expressions within methods. Some methods will take different execution paths based upon whether the supplied input was an array or a string:

```ion
echo $len(@array) # Prints number of elements in array
echo $len($string) # Prints the number of graphemes in string
echo $split($(cmd), ",") # Specialized splitting of output of the command
```

They are pretty powerful features. But we do need to improve the parsing of these methods a bit more, and add some more builtin methods.

**Comment by BafDyce** [comment]
@mmstick Thanks for the explanation!

I somehow missed that feature completely in the docs. Is it not explained there or did I just not see that chapter?

These methods indeed look like a really powerful feature, I think I need to study them in more detail.

**Comment by mmstick** [comment]
It's not documented in the mdbook, but there is an issue that highlights what methods are implemented / awaiting to be implemented, and open for discussion on other method ideas and implementation details. #441

**Comment by xc3r0x** [comment]
But isn't it saver to disable the plugins by default?

**Comment by mmstick** [comment]
Root is now disallowed from executing plugins, and namespace plugins are disabled by default.
